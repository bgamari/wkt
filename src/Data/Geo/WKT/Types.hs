{-# LANGUAGE TemplateHaskell, DeriveDataTypeable, DeriveGeneric #-}

module Data.Geo.WKT.Types where

import GHC.Generics
import Data.Typeable
import Control.Lens
import Linear

-- ^ From http://www.geoapi.org/3.0/javadoc/org/opengis/referencing/doc-files/WKT.html

data Parameter = Parameter { _paramName     :: String
                           , _paramValue    :: Double
                           }
               deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''Parameter

data Authority = Authority { _authorityName :: String
                           , _authorityCode :: String
                           }
               deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''Authority

data Unit = Unit { _unitName               :: String
                 , _unitConversionFactor   :: Double
                 , _unitAuthority          :: Maybe Authority
                 }
          deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''Unit

data Spheroid = Spheroid { _spheroidName                :: String
                         , _spheroidSemiMajor           :: Double
                         , _spheroidInverseFlattening   :: Double
                         , _spheroidAuthority           :: Maybe Authority
                         }
              deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''Spheroid

data ToWGS84 = ToWGS84 { _towgs84D      :: V3 Double
                       , _towgs84E      :: V3 Double
                       , _towgs84PPM    :: Double
                       }
             deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''ToWGS84

data Datum = Datum { _datumName         :: String
                   , _datumSpheroid     :: Spheroid
                   , _datumToWGS84      :: Maybe ToWGS84
                   , _datumAuthority    :: Maybe Authority
                   }
           deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''Datum

data PrimeMeridian = PrimeMeridian { _primemName      :: String
                                   , _primemLongitude :: Double
                                   , _primemAuthority :: Maybe Authority
                                   }
                   deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''PrimeMeridian     

data AxisDirection = North | South | East | West | Up | Down | Other
                   deriving (Show, Eq, Ord, Typeable, Generic)

data Axis = Axis { _axisName      :: String
                 , _axisDirection :: AxisDirection
                 }
          deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''Axis     

data GeographicCS = GeogCS { _geogcsName            :: String
                           , _geogcsDatum           :: Datum
                           , _geogcsPrimeMeridian   :: PrimeMeridian
                           , _geogcsAngularUnit     :: Unit
                           , _geogcsTwinAxes        :: Maybe (Axis, Axis)
                           , _geogcsAuthority       :: Maybe Authority
                           }
                  deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''GeographicCS

data Projection = Proj { _projName         :: String
                       , _projAuthority    :: Maybe Authority
                       }
                deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''Projection

data ProjectedCS = ProjCS { _projcsName             :: String
                          , _projcsGeographicCS     :: GeographicCS
                          , _projcsProjection       :: Projection
                          , _projcsParameters       :: [Parameter]
                          , _projcsLinearUnit       :: Unit
                          , _projcsTwinAxes         :: Maybe (Axis, Axis)
                          , _projcsAuthority        :: Maybe Authority
                          }
                 deriving (Show, Eq, Ord, Typeable, Generic)
makeLenses ''ProjectedCS
