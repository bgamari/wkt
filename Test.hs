import Data.Geo.WKT
import Text.Trifecta
import Test.Tasty
import Test.Tasty.Golden
import System.FilePath

testVsFile :: String -> TestTree
testVsFile name =
    goldenVsFile name expectedFile outFile run
  where
    expectedFile = "tests"</>name<.>"expected"
    outFile      = "tests"</>name<.>"out"
    run = parseFromFile projectedCS ("tests"</>name) >>= writeFile outFile . show

main :: IO ()
main = defaultMain $ testGroup "tests"
    [ testVsFile "test.prj"
    ]
