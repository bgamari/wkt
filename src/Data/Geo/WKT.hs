module Data.Geo.WKT ( module Data.Geo.WKT.Types
                    , module Data.Geo.WKT.Parser
                    ) where

import Data.Geo.WKT.Types
import Data.Geo.WKT.Parser

