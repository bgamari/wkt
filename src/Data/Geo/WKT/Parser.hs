module Data.Geo.WKT.Parser where

import Data.Geo.WKT.Types
import Linear
import Control.Applicative
import Text.ParserCombinators.Parsec hiding (many)
import Text.ParserCombinators.Parsec.Number hiding (number)

object :: String -> Parser a -> Parser a
object keyword parser = do
    string keyword
    between (char '[') (char ']') parser
   
quotedString :: Parser String
quotedString = do
    char '"'
    manyTill anyChar (char '"')

fieldSep :: Parser ()
fieldSep = char ',' >> spaces

number :: Parser Double
number = floating2 True

twinAxes = do
    ax1 <- axis
    fieldSep 
    ax2 <- axis
    fieldSep
    return (ax1, ax2)

unit :: Parser Unit
unit = object "UNIT" $ do
    name <- quotedString
    fieldSep
    conv <- number
    auth <- optionMaybe $ fieldSep *> authority
    return $ Unit name conv auth
    
parameter :: Parser Parameter
parameter = object "PARAMETER" $ do
    name <- quotedString
    fieldSep
    value <- number
    return $ Parameter name value 

authority :: Parser Authority
authority = object "AUTHORITY" $ do
    name <- quotedString
    fieldSep
    code <- quotedString
    return $ Authority name code

axis :: Parser Axis
axis = object "AXIS" $ do
    name <- quotedString
    fieldSep
    dir <- choice [ string "NORTH" >> return North
                  , string "SOUTH" >> return South
                  , string "EAST"  >> return East
                  , string "WEST"  >> return West
                  , string "UP"    >> return Up
                  , string "DOWN"  >> return Down
                  , string "OTHER" >> return Other
                  ]
    return $ Axis name dir

projection :: Parser Projection
projection = object "PROJECTION" $ do
    name <- quotedString
    auth <- optionMaybe $ fieldSep *> authority
    return $ Proj name auth

spheroid :: Parser Spheroid
spheroid = object "SPHEROID" $ do
    name <- quotedString
    fieldSep
    semiMajor <- number
    fieldSep
    invFlat <- number
    auth <- optionMaybe $ fieldSep *> authority
    return $ Spheroid name semiMajor invFlat auth
     
datum :: Parser Datum
datum = object "DATUM" $ do             
    name <- quotedString
    fieldSep
    s <- spheroid
    wgs <- optionMaybe $ fieldSep *> toWGS84
    auth <- optionMaybe $ fieldSep *> authority
    return $ Datum name s wgs auth

primeMeridian :: Parser PrimeMeridian
primeMeridian = object "PRIMEM" $ do
    name <- quotedString
    fieldSep
    long <- number
    auth <- optionMaybe $ fieldSep *> authority
    return $ PrimeMeridian name long auth
    
toWGS84 :: Parser ToWGS84
toWGS84 = object "TOWGS84" $ do
    d <- V3 <$> number <*> number <*> number        
    e <- V3 <$> number <*> number <*> number        
    ppm <- number
    return $ ToWGS84 d e ppm

projectedCS :: Parser ProjectedCS
projectedCS = object "PROJCS" $ do
    name <- quotedString
    fieldSep
    geogcs <- geographicCS
    fieldSep
    proj <- projection
    fieldSep
    params <- many $ parameter <* fieldSep
    linearUnit <- unit
    axes <- optionMaybe $ fieldSep *> twinAxes
    auth <- optionMaybe $ fieldSep *> authority
    return $ ProjCS name geogcs proj params linearUnit axes auth
    
geographicCS :: Parser GeographicCS
geographicCS = object "GEOGCS" $ do
    name <- quotedString
    fieldSep
    dat <- datum
    fieldSep
    primem <- primeMeridian
    fieldSep
    angularUnit <- unit
    axes <- optionMaybe $ fieldSep *> twinAxes
    auth <- optionMaybe $ fieldSep *> authority
    return $ GeogCS name dat primem angularUnit axes auth
    
