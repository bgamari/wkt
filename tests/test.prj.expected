Just (ProjCS {_projcsName = "NAD_1983_StatePlane_Massachusetts_Mainland_FIPS_2001_Feet", _projcsGeographicCS = GeogCS {_geogcsName = "GCS_North_American_1983", _geogcsDatum = Datum {_datumName = "D_North_American_1983", _datumSpheroid = Spheroid {_spheroidName = "GRS_1980", _spheroidSemiMajor = 6378137.0, _spheroidInverseFlattening = 298.257222101, _spheroidAuthority = Nothing}, _datumToWGS84 = Nothing, _datumAuthority = Nothing}, _geogcsPrimeMeridian = PrimeMeridian {_primemName = "Greenwich", _primemLongitude = 0.0, _primemAuthority = Nothing}, _geogcsAngularUnit = Unit {_unitName = "Degree", _unitConversionFactor = 1.74532925199433e-2, _unitAuthority = Nothing}, _geogcsTwinAxes = Nothing, _geogcsAuthority = Nothing}, _projcsProjection = Proj {_projName = "Lambert_Conformal_Conic", _projAuthority = Nothing}, _projcsParameters = [Parameter {_paramName = "False_Easting", _paramValue = 656166.6666666665},Parameter {_paramName = "False_Northing", _paramValue = 2460625.0},Parameter {_paramName = "Central_Meridian", _paramValue = -71.5},Parameter {_paramName = "Standard_Parallel_1", _paramValue = 41.71666666666667},Parameter {_paramName = "Standard_Parallel_2", _paramValue = 42.68333333333333},Parameter {_paramName = "Latitude_Of_Origin", _paramValue = 41.0}], _projcsLinearUnit = Unit {_unitName = "Foot_US", _unitConversionFactor = 0.3048006096012192, _unitAuthority = Nothing}, _projcsTwinAxes = Nothing, _projcsAuthority = Nothing})